contributions/views.py

```diff
@@ -1,158 +1,266 @@
+        """
+        Kick off an STK Push for a member’s scheduled contribution.
+        """
+        user = request.user
+        schedule_id = request.data.get('schedule')
+        amount      = request.data.get('amount')
+
+        # Validate schedule & membership
+        try:
+            schedule_uuid = uuid.UUID(schedule_id)
+            schedule = ContributionSchedule.objects.get(id=schedule_uuid)
+        except (ValueError, DjangoValidationError, ContributionSchedule.DoesNotExist):
+            raise ValidationError({'schedule': 'Invalid schedule ID.'})
+        except Membership.DoesNotExist:
+            raise PermissionDenied('Not a member of this Chama.')
+        
+
+        # verify membership
+        try:
+            Membership.objects.get(user=user, chama=schedule.chama)
+        except Membership.DoesNotExist:
+            raise PermissionDenied('Not a member of this Chama.')
+
+        # Build callback URL for Daraja to hit
+        callback_url = request.build_absolute_uri(reverse('mpesa-callback'))
+
+        # Use schedule ID + timestamp as reference
+        reference = f"{schedule_id}-{int(timezone.now().timestamp())}"
+
+        # STK Push
+        try:
+            resp = MpesaDarajaClient.initiate_stk_push(
+                phone_number=user.phone,
+                amount=amount,
+                reference=reference,
+                callback_url=callback_url
             )
+        except requests.HTTPError as e:
+            raise ValidationError({'mpesa': 'Failed to initiate STK Push.'})
+
+        return Response({
+            'CheckoutRequestID': resp.get('CheckoutRequestID'),
+            'ResponseCode':      resp.get('ResponseCode'),
+            'ResponseDescription': resp.get('ResponseDescription'),
+            'MerchantRequestID': resp.get('MerchantRequestID'),
+            'reference':         reference
+        }, status=200)
+
+@api_view(['POST'])  # must be outermost
+@authentication_classes([])  # disables auth
+@permission_classes([])  # disables permission checks
+@csrf_exempt  # optional, but helpful for external services
+def mpesa_callback(request):
+    """
+    Endpoint for M-Pesa payment confirmations.
+    """
+    print(">>> mpesa_callback view HIT")
+    required = ['phone', 'amount', 'reference', 'chama_id', 'schedule_id']
+    missing  = [f for f in required if not request.data.get(f)]
+    if missing:
+        return Response(
+            {'error': f"Missing required fields: {', '.join(missing)}"},
+            status=status.HTTP_400_BAD_REQUEST
+        )
+
+    phone     = request.data['phone']
+    reference = request.data['reference'].strip()
+    amount    = request.data['amount']
+    schedule_id = request.data['schedule_id']
+    chama_id  = request.data['chama_id']
+
+    # lookup user & membership
+    from users.models import User
+    try:
+        user = User.objects.get(phone=phone)
+    except User.DoesNotExist:
+        raise NotFound("User not found.")
```

Comment:
This doesn't account for users who signed up with email or those using a different phone number for payments. The same applies to the phone number used for initiating the STK push, it may not always match the one on their profile.